cmd/azure-config-credentials-injector/main.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	clientIDEnvKey     = "AZURE_CLIENT_ID"
+	clientSecretEnvKey = "AZURE_CLIENT_SECRET"
+
+	clientIDCloudConfigKey     = "AADClientID"
+	clientSecretCloudConfigKey = "AADClientSecret"
+)
+
+func main() {
+	rootCmd := &cobra.Command{
+		Use:   "azure-config-credentials-injector [cloud-config-path] [output-path]",
+		Short: "Cloud config credentials injection tool for azure cloud platform",
+		Args:  cobra.ExactArgs(2),
+		RunE:  mergeCloudConfig,
+	}
+
+	if err := rootCmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func mergeCloudConfig(_ *cobra.Command, args []string) error {
+	cloudConfigPath := args[0]
+	if _, err := os.Stat(cloudConfigPath); os.IsNotExist(err) {
+		return err
+	}
+
+	azureClientId, found := os.LookupEnv(clientIDEnvKey)
+	if !found {
+		return fmt.Errorf("%s env variable should be set up", clientIDEnvKey)
+	}
+
+	azureClientSecret, found := os.LookupEnv(clientSecretEnvKey)
+	if !found {
+		return fmt.Errorf("%s env variable should be set up", clientSecretEnvKey)
+	}
+
+	cloudConfig, err := readCloudConfig(cloudConfigPath)
+	if err != nil {
+		return err
+	}
+
+	preparedCloudConfig, err := prepareCloudConfig(cloudConfig, azureClientId, azureClientSecret)
+	if err != nil {
+		return err
+	}
+
+	outputPath := args[1]
+	if err := writeCloudConfig(outputPath, preparedCloudConfig); err != nil {
+		return err
+	}
```

Comment:
i'm curious about these failure conditions, do we not need a log message?

Reply:
It would fail in case if file path is not valid (e.g. directory), permissions issues and so on. Error would be propagated to toplevel and printed.

Reply:
got it, thanks!

Reply:
Could you expand the errors here using `fmt.Errorf` to provide additonal detail about what exactly is failing, will almost certainly make it easier to find the issue when you can tie an individual error message to a particular if statement